### ISSUE TITLE
attempt to index global 'process' (a nil value)

### ISSUE THREAD

**Issue by joerg-krause** [luvit/luvit#948]
Running a luvi app that layers on top of luvit (version 1.12.1) using `fs.createReadStream()` fails:

```
# luvi --version
luvi version: v2.7.5
zlib version: 1.2.11
rex version: 8.40 2017-01-11
libuv version: 1.11.0
ssl version: OpenSSL 1.0.2k  26 Jan 2017, lua-openssl 0.5.1

# /usr/bin/luvi /path/to/app /path/to/luvit
[string "bundle:/deps/stream/stream_readable.lua"]:805: attempt to index global 'process' (a nil value)
```

Minimal luvit app to reproduce:

``` Lua
local fs = require("fs")
local s = fs.createReadStream("/tmp/test.sock")
s:on("data", function (chunk) end)
```

**Comment by creationix** [comment]
The stream module probably shouldn't depend on `_G.process` from luvit core.

But I'm curious, how are you bootstrapping your luvi app?  The luvit main function is supposed to setup the globals for custom luvi apps. https://github.com/luvit/luvit/blob/master/examples/custom-binary/main.lua

**Comment by joerg-krause** [comment]
I'm building luvi using [Buildroot](https://git.buildroot.net/buildroot/tree/package/luvi/luvi.mk) and I fetch luvit from the lit server: `curl https://lit.luvit.io/packages/luvit/luvit/latest.zip > luvit.zip && unzip luvit.zip -d luvit`.

luvits main function is not called when used as a bundle, but init.lua should, right?

**Comment by creationix** [comment]
Running luvit from unzipped tree should be the same as from zip directly.  I'm not sure what exactly you're doing.

**Comment by joerg-krause** [comment]
I am cross-compiling luvi using Buildroot for my ARM board. On the board I am running the cross-compiled luvi: `luvi /app /luvit` whereas `/app` is the is the path to the `main.lua` and `/luvit` is the path to the unzipped luvit app. The main.lua is the minimal luvit app to reproduce the issue as quoted above. The unzipped luvit app is fetched from the lit server as stated above.

In short, I am doing what is written in the luvi README:
```
# Run an app that layers on top of luvit
  luvi path/to/app path/to/luvit
```

I was doing this track down an [issue](https://github.com/luvit/luvit/issues/940#issuecomment-282231500) with the latest luvit version.

**Comment by squeek502** [comment]
I believe that is an outdated usage of luvi. Either of the following methods should work, but the second seems like it fits your use case better.

**Method 1**

Wrap your `main.lua` in the following:
```
return require('luvit')(function (...)
  -- your stuff goes here
end, ...)
```
and run `lit install luvit/luvit` inside `/app` (or unzip luvit into `/app/deps`, but make sure to create the subdir `/app/deps/luvit/` and add Luvit's `init.lua` to it).

Then, doing `luvi /app` should work as expected. This is the method @creationix [linked to an example of earlier](https://github.com/luvit/luvit/blob/master/examples/custom-binary/main.lua)

**Method 2**

Use `luvi /path/to/luvit -- /path/to/app` instead, which will act exactly like `luvit /path/to/app` would. This is [how its suggested to run the Luvit tests using a custom luvi](https://github.com/luvit/luvit#notes-to-maintainers)

**Comment by joerg-krause** [comment]
@squeek502 I see!

I can confirm that method 2 works, if I use it the following way: `/usr/bin/luvi -- /root/luvit /root/app/main.lua`.

@creationix Looks like the luvi help needs an update?

However, method 1 does not work when unzipping luvit's lit package into `/app/deps`:
```
# luvi ./app
[string "bundle:main.lua"]:1: module 'luvit' not found:
	no field package.preload['luvit']
	no file './luvit.lua'
	no file '/usr/share/luajit-2.0.4/luvit.lua'
	no file '/usr/local/share/lua/5.1/luvit.lua'
	no file '/usr/local/share/lua/5.1/luvit/init.lua'
	no file '/usr/share/lua/5.1/luvit.lua'
	no file '/usr/share/lua/5.1/luvit/init.lua'
	no file './luvit.so'
	no file '/usr/local/lib/lua/5.1/luvit.so'
	no file '/usr/lib/lua/5.1/luvit.so'
	no file '/usr/local/lib/lua/5.1/loadall.so'
# ls ./app/
deps      main.lua
# ls ./app/deps/
luvit
# ls ./app/deps/luvit/
deps         init.lua     main.lua     package.lua
```

Unfortunately, I am not able to test using lit. Running busybox' unzip on my ARM board is not able to unzip the lit package from the lit server. It breaks with: `unzip: corrupted data`. However, unzipping works on my host Linux system using UnZip. I guess, this is a bug in busybox' unzip.

**Comment by squeek502** [comment]
Ah, `lit` installs `luvit` into `deps` differently. It should be:

```
app
+-- deps
    +-- http.lua
    +-- (etc, all luvit deps should be installed in the root of deps),
    +-- luvit
         +-- init.lua
         +-- main.lua
         +-- package.lua
```

I believe Method 1 is not working for you now, even though you have luvit's `init.lua` in the right spot, because `luvi` will bootstrap `deps/require.lua` automatically if it finds it, but you have `require.lua` in `deps/luvit/deps` so it doesn't find it, and therefore it's looking for `luvit` using the normal Lua require, which fails.

**Comment by creationix** [comment]
Also what's probably missing from that example, is luvi used to auto-detect and bootstrap `deps/require.lua` if it existed.  I'm not sure this is the case anymore.  Look at luvit's main to see how it bootstraps currently for latest info on the luvi it uses.

**Comment by creationix** [comment]
Also, thanks @squeek502 for helping!  My whole family has the flu and I'm a bit under the weather.  Also expecting our 5th child any day now.

**Comment by joerg-krause** [comment]
@creationix Get well soon! And all the best for the forthcoming birth!

@squeek502 Thanks for the help! I am fine with method 1 working for me.

I'll close this for now as it turned out that I used an outdated usage of luvi. I wil open a new issue instead!

**Comment by joerg-krause** [comment]
> The stream module probably shouldn't depend on _G.process from luvit core.

@creationix You're right. Running `luvi /path/to/app /path/to/luvit` only fails for the stream module. This has nothing to do with an outdated usage of luvi.

Replacing `process.nextTick` by `timer.setImmediate` resolved the issue for me.
